### PR TITLE
Prefer stack storage for highly-contended temporary array

### PR DESCRIPTION
**Pull request by sharwell**
Fixes [AB#1449857](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1449857)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs [hunk 1]
```diff
@@ -1311,19 +1312,16 @@ private BetterResult MoreSpecificOperator(BinaryOperatorSignature op1, BinaryOpe
                 op2Right = op2.RightType;
             }
 
-            var uninst1 = ArrayBuilder<TypeSymbol>.GetInstance();
-            var uninst2 = ArrayBuilder<TypeSymbol>.GetInstance();
+            using var uninst1 = TemporaryArray<TypeSymbol>.Empty;
+            using var uninst2 = TemporaryArray<TypeSymbol>.Empty;
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs:1316]: 📝 `TemporaryArray<T>` stores up to 4 elements on the stack, and only uses `ArrayBuilder<T>.GetInstance()` for 5 or more elements.

**Comment by genlu** [comment]
Hey @sharwell, do you mind explaining what did you see in the trace prompted this change? Thanks!

**Comment by sharwell** [comment]
@genlu The trace shows high CPU usage in ``ObjectPool`1[System.__Canon].Allocate()``. I've seen this in the past on my 2990WX, and has to do with `Interlocked.CompareExchange` needing to synchronize the caches of all 64 processors.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs:1316]: Being pinned to a core might make sense as well.

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs:1316]: There are options there, but it's difficult to implement them without penalizing some other scenario. Stack-based allocations have done a good job of addressing this situation in IDE code, so I would expect them to do the same here.

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for second review

**Review by jaredpar — APPROVED** [review]

**Pull request merged**